examples/async-sentiment.rs

```diff
@@ -0,0 +1,63 @@
+use std::{
+    sync::mpsc,
+    thread::{self, JoinHandle},
+};
+
+use anyhow::Result;
+use rust_bert::pipelines::sentiment::{Sentiment, SentimentConfig, SentimentModel};
+use tokio::{sync::oneshot, task};
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    let (_handle, classifier) = SentimentClassifier::spawn();
+
+    let texts = vec![
+        "Classify this positive text".to_owned(),
+        "Classify this negative text".to_owned(),
+    ];
+    let sentiments = classifier.predict(texts).await?;
+    println!("Results: {:?}", sentiments);
+
+    Ok(())
+}
+
+/// Message type for internal channel, passing around texts and return value
+/// senders
+type Message = (Vec<String>, oneshot::Sender<Vec<Sentiment>>);
+
+/// Runner for sentiment classification
+#[derive(Debug, Clone)]
+pub struct SentimentClassifier {
+    sender: mpsc::SyncSender<Message>,
+}
+
+impl SentimentClassifier {
+    /// Spawn a classifier on a separate thread and return a classifier instance
+    /// to interact with it
+    pub fn spawn() -> (JoinHandle<Result<()>>, SentimentClassifier) {
+        let (sender, receiver) = mpsc::sync_channel(100);
+        let handle = thread::spawn(move || Self::runner(receiver));
+        (handle, SentimentClassifier { sender })
+    }
+
+    /// The classification runner itself
+    fn runner(receiver: mpsc::Receiver<Message>) -> Result<()> {
+        // Needs to be in sync runtime, async doesn't work
+        let model = SentimentModel::new(SentimentConfig::default())?;
+
+        while let Ok((texts, sender)) = receiver.recv() {
+            let texts: Vec<&str> = texts.iter().map(String::as_str).collect();
+            let sentiments = model.predict(texts);
+            sender.send(sentiments).expect("sending results");
+        }
+
+        Ok(())
+    }
+
+    /// Make the runner predict a sample and return the result
+    pub async fn predict(&self, texts: Vec<String>) -> Result<Vec<Sentiment>> {
+        let (sender, receiver) = oneshot::channel();
+        task::block_in_place(|| self.sender.send((texts, sender)))?;
```

Comment:
```suggestion
       self.sender.send((texts, sender))?;
```
I believe blocking here is not necessary - the model creation needs to be run in a blocking context - once created the predict methods can be non-blocking. The suggested code compiles and runs fine on my side

Reply:
I am not sure if I understand correctly, these are two points.

While running the prediction in the async runtim is possible, it is not recommended due extensive CPU usage for a long time without yielding to the scheduler. Therefore, running everything model-related in a separate thread might be the best solution.

Regarding the `block_in_place` here: Unbounded channels as created using `mpsc::channel()` would not block on sending messages and you would therefore be right. Unfortunately, this kind of `mpsc::Sender` does not implement Sync and cannot be shared across threads / used in general tokio runtimes.
The solution is to use `mpsc::sync_channel()`, which returns a `mpsc::SyncSender`. This one however has a maximum amount of messages in the queue and blocks execution whenever the channel buffer is full. Consequently, the block_in_place, but not `spawn_blocking`.